src/lib/structures/objective.ml

```diff
@@ -0,0 +1,134 @@
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module Function = struct
+  type t = {
+    e : Expr.t;
+    is_max : bool;
+  }
+
+  let equal { e = e1; is_max = m1 } { e = e2; is_max = m2 } =
+    Bool.equal m1 m2 && Expr.equal e1 e2
+
+  let mk ~is_max e = { e; is_max }
+
+  let pp ppf { e; _ } = Expr.print ppf e
+end
+
+module Value = struct
+  type limit_kind =
+    | Above
+    | Below
+
+  type t =
+    | Minfinity
+    | Pinfinity
+    | Value of Expr.t
+    | Limit of limit_kind * Expr.t
+    | Unknown
+
+  let pp ppf v =
+    match v with
+    | Minfinity -> Fmt.pf ppf "(- oo)"
+    | Pinfinity -> Fmt.pf ppf "(+ oo)"
+    | Value w -> Expr.print ppf w
+    | Limit (Above, w) -> Fmt.pf ppf "(+ %a epsilon)" Expr.print w
+    | Limit (Below, w) -> Fmt.pf ppf "(- %a epsilon)" Expr.print w
+    | Unknown -> Fmt.pf ppf "(interval (- oo) (+ oo))"
+end
+
+module Model = struct
+  type s = { f: Function.t; order: int }
+
+  module M =  Map.Make (struct
+      type t = s
+
+      let compare { order = o1; _ } { order = o2; _ } = o1 - o2
+    end)
+
+  type t = Value.t M.t
+
+  let empty = M.empty
+  let is_empty = M.is_empty
+  let fold g = M.fold (fun { f; _ } acc -> g f acc)
+
+  let add o v mdl =
+    let order =
+      match M.find_first (fun { f; _ } -> Function.equal f o) mdl with
+      | ({ order; _ }, _) -> order
+      | exception Not_found -> M.cardinal mdl
+    in
+    M.add { f = o; order } v mdl
+
+  let pp_binding ppf ({ f; _ }, v) =
+    Fmt.pf ppf "(%a %a)" Function.pp f Value.pp v
+
+  let pp ppf mdl =
+    if M.is_empty mdl then
+      Fmt.pf ppf "@[<v 2>(objectives @]@,)"
+    else
+      Fmt.pf ppf "@[<v 2>(objectives @,%a@]@,)"
+        (Fmt.iter_bindings ~sep:Fmt.cut M.iter pp_binding) mdl
+
+  let functions mdl =
+    M.bindings mdl
+    |> List.map (fun ({ f; _ }, _) -> f)
+
+  let has_no_limit mdl =
+    M.for_all
+      (fun _ v ->
+         match (v : Value.t) with
+         | Pinfinity | Minfinity | Limit _ -> false
+         | Value _ | Unknown -> true
+      ) mdl
+
+  exception Found of Function.t * int
+
+  let next_unknown ~for_model mdl =
+    try
+      M.iter (fun { f; order } v ->
+          match (v : Value.t) with
+          | Value _ -> ()
+          | Limit _ | Pinfinity | Minfinity when for_model -> raise Exit
+          | _ -> raise (Found (f, order))
+        ) mdl;
+      None
+    with
+    | Found (f, order) -> Some (f, order)
+    | Exit -> None
+
+  let reset_until mdl o =
+    M.fold
+      (fun { f; order } v acc ->
+         if order < o then M.add { f; order } v acc
```

Comment:
Nit: this could just take `mdl` as the initial value for `acc` and not do anything in this case.